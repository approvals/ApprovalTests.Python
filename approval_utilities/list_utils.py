from typing import Any, Callable, List, Optional


def format_list(alist: List[Any], formatter: Optional[Callable], header: str) -> str:
    if formatter is None:
        formatter = FormatLineItem().print_item
    text = (header + "\n\n") if header else ""
    for i in alist:
        text += formatter(i) + "\n"
    return text


class FormatLineItem:
    def __init__(self) -> None:
        self.index = 0

    def print_item(self, item: Any) -> str:
        text = str(self.index) + ") " + str(item)
        self.index += 1
        return text
