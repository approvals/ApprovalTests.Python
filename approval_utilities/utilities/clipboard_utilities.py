def copy_to_clipboard(text: str) -> None:
    """
    This acts as a wrapper to defer the importing of pyperclip util actual usage
    """
    import pyperclip

    pyperclip.copy(text)
