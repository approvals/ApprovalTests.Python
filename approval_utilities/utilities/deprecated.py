import functools
import warnings
from typing import Callable


def deprecated(reason: str) -> Callable:
    def decorator(func: Callable) -> Callable:
        """This is a decorator which can be used to mark functions
        as deprecated. It will result in a warning being emitted
        when the function is used."""

        @functools.wraps(func)
        def new_func(*args, **kwargs):
            warnings.simplefilter("always", DeprecationWarning)  # turn off filter
            warnings.warn(
                f"Call to deprecated function {func.__name__}. {reason}",
                category=DeprecationWarning,
                stacklevel=2,
            )
            warnings.simplefilter("default", DeprecationWarning)  # reset filter
            return func(*args, **kwargs)

        return new_func

    return decorator
