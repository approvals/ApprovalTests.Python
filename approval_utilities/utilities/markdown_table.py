from typing import TYPE_CHECKING, Any, Callable, Iterable

from typing_extensions import override

from approval_utilities.approvaltests.core.verifiable import Verifiable
from approval_utilities.approvaltests.core.verify_parameters import VerifyParameters

if TYPE_CHECKING:
    from approvaltests.core.options import Options


class MarkdownTable(Verifiable):
    def __init__(self) -> None:
        self.markdown: str = ""

    @override
    def get_verify_parameters(self, options: "Options") -> VerifyParameters:
        return VerifyParameters(options.for_file.with_extension(".md"))

    @staticmethod
    def with_headers(*column_names: str) -> "MarkdownTable":
        table = MarkdownTable()
        table.add_rows(*column_names)
        dividers = map(lambda _: "---", column_names)
        table.add_rows(*dividers)
        return table

    def add_rows(self, *column_names: str) -> "MarkdownTable":
        self.markdown += MarkdownTable.print_row(*column_names)
        return self

    @override
    def __str__(self) -> str:
        return self.markdown

    @staticmethod
    def print_row(*column_names: Any) -> str:
        row = "| " + " | ".join(map(str, column_names)) + " |"
        return row + "\n"

    def add_rows_for_inputs(
        self, inputs: Iterable[Any], *input_transformers: Callable[[Any], Any]
    ) -> "MarkdownTable":
        def transform_resolver_for_input(
            input_value: Any,
        ) -> Callable[[Callable[[Any], Any]], Any]:
            return lambda transform: transform(input_value)

        for row_input in inputs:
            row = [row_input]
            row += map(transform_resolver_for_input(row_input), input_transformers)
            self.add_rows(*row)
        return self
