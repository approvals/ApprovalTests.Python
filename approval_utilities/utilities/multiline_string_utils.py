import textwrap
from typing import Optional


def remove_indentation_from(text: Optional[str]) -> str:
    if not text:
        return ""
    cleaned = textwrap.dedent(text + "|")
    if cleaned.startswith("\n"):
        cleaned = cleaned[1:]
    return cleaned[:-1]
