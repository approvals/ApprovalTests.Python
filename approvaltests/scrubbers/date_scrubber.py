from typing import List, Tuple

from approvaltests.scrubbers import create_regex_scrubber
from approvaltests.scrubbers.scrubbers import Scrubber


class DateScrubber:
    @staticmethod
    def get_supported_formats() -> List[Tuple[str, List[str]]]:
        return [
            (
                "[a-zA-Z]{3} [a-zA-Z]{3} \\d{2} \\d{2}:\\d{2}:\\d{2}",
                ["Tue May 13 16:30:00"],
            ),
            (
                "[a-zA-Z]{3} [a-zA-Z]{3} \\d{2} \\d{2}:\\d{2}:\\d{2} [a-zA-Z]{3,4} \\d{4}",
                ["Wed Nov 17 22:28:33 EET 2021"],
            ),
            (
                "[a-zA-Z]{3} [a-zA-Z]{3} \\d{2} \\d{4} \\d{2}:\\d{2}:\\d{2}.\\d{3}",
                ["Tue May 13 2014 23:30:00.789"],
            ),
            (
                "[a-zA-Z]{3} [a-zA-Z]{3} \\d{2} \\d{2}:\\d{2}:\\d{2} -\\d{4} \\d{4}",
                ["Tue May 13 16:30:00 -0800 2014"],
            ),
            (
                "\\d{2} [a-zA-Z]{3} \\d{4} \\d{2}:\\d{2}:\\d{2},\\d{3}",
                ["13 May 2014 23:50:49,999"],
            ),
            (
                "[a-zA-Z]{3} \\d{2}, \\d{4} \\d{2}:\\d{2}:\\d{2} [a-zA-Z]{2} [a-zA-Z]{3}",
                ["May 13, 2014 11:30:00 PM PST"],
            ),
            ("\\d{2}:\\d{2}:\\d{2}", ["23:30:00"]),
            (
                "\\d{4}/\\d{2}/\\d{2} \\d{2}:\\d{2}:\\d{2}.\\d{2}\\d",
                ["2014/05/13 16:30:59.786"],
            ),
            (
                "\\d{4}-\\d{1,2}-\\d{1,2}T\\d{1,2}:\\d{2}Z",
                [
                    "2020-9-10T08:07Z",
                    "2020-09-9T08:07Z",
                    "2020-09-10T8:07Z",
                    "2020-09-10T08:07Z",
                ],
            ),
            (
                "\\d{4}-\\d{1,2}-\\d{1,2}T\\d{1,2}:\\d{2}:\\d{2}Z",
                ["2020-09-10T08:07:89Z"],
            ),
            (
                "\\d{4}-\\d{1,2}-\\d{1,2}T\\d{1,2}:\\d{2}\\:\\d{2}\\.\\d{3}Z",
                ["2020-09-10T01:23:45.678Z"],
            ),
            (
                r"\d{4}-\d{1,2}-\d{1,2}(?:T| )\d{1,2}:\d{2}:\d{2}\.\d{6}",
                ["2023-07-16 17:39:03.293919", "2023-12-06T11:59:47.090226"],
            ),
            ("\\d{8}T\\d{6}Z", ["20210505T091112Z"]),
            (
                r"(Mon|Tue|Wed|Thu|Fri|Sat|Sun)\s(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s([0-3]?\d)\s([0-1]\d:[0-5]\d:[0-5]\d)\s(\d{4})",
                ["Tue May 13 16:30:00 2014", "Wed Dec 11 14:59:44 2024"],
            ),
            (
                r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\+\d{2}:\d{2}",
                ["2021-09-10T08:07:00+03:00", "2021-01-01T00:00:00+00:00"],
            ),
            (
                r"[12]\d{3}[01]\d[0-3]\d_[0-2]\d[0-5]\d[0-5]\d",
                ["20250527_125703"],
            ),
        ]

    def __init__(self, date_regex: str):
        self.date_regex = date_regex

    def scrub(self, date_str: str) -> str:
        return create_regex_scrubber(self.date_regex, lambda t: f"<date{t}>")(date_str)

    @staticmethod
    def get_scrubber_for(example: str) -> Scrubber:
        supported = ""
        for date_regex, examples in DateScrubber.get_supported_formats():
            supported += f"    {examples[0]} | {date_regex} \n"
            scrubber = DateScrubber(date_regex)
            if scrubber.scrub(example) == "<date0>":
                return scrubber.scrub

        raise Exception(
            f"No match found for '{example}'.\n Feel free to add your date at https://github.com/approvals/ApprovalTests.Python/issues/124 \n Current supported formats are: \n{supported}"
        )
