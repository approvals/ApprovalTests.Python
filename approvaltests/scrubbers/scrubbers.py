import re
from collections import abc, defaultdict
from typing import Callable, DefaultDict, Union

from approval_utilities.utilities.logger.logging_instance import print_type

Scrubber = Callable[[str], str]


# def print_type(value):
#     return f"<{type(value).__name__}>"
def create_regex_scrubber(
    regex: str, function_or_replace_string: Union[Callable[[int], str], str]
) -> Scrubber:
    if isinstance(function_or_replace_string, str):
        return lambda t: _replace_regex(t, regex, lambda _: function_or_replace_string)
    elif callable(function_or_replace_string):
        return lambda t: _replace_regex(t, regex, function_or_replace_string)
    else:
        raise TypeError(
            f"Parameter function_or_replace_string expects a string or callable, but got {print_type(function_or_replace_string)}"
        )


def _replace_regex(text: str, regex: str, replacement: Callable[[int], str]) -> str:
    matches: DefaultDict[str, int] = defaultdict(lambda: len(matches))
    return re.sub(regex, lambda m: replacement(matches[m.group(0)]), text)


def scrub_all_dates(date: str) -> str:
    return create_regex_scrubber(
        r"\d{4}-\d{2}-\d{2}\ \d{2}:\d{2}:\d{2}", lambda t: f"<date{t}>"
    )(date)


def scrub_all_guids(data: str) -> str:
    return create_regex_scrubber(
        r"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}",
        lambda t: f"<guid_{t}>",
    )(data)


def combine_scrubbers(*scrubbers: Scrubber) -> Scrubber:
    def combined(data: str) -> str:
        for scrubber in scrubbers:
            data = scrubber(data)
        return data

    return combined


def templates_regex_scrubber_with_lambda() -> Scrubber:
    """
    This method exists as a convenient way to get an example scrubber for you to use.
    To use this template, simply inline the method in your IDE.
    """
    return create_regex_scrubber(
        "your pattern here: [a-zA-Z]+/d{4}", lambda t: f"<your replacement_{t}>"
    )


def templates_regex_scrubber_with_replacement() -> Scrubber:
    """
    This method exists as a convient way to get an example scrubber for you to use.
    To use this template, simply inline the method in your IDE.
    """
    return create_regex_scrubber(
        "your pattern here: [a-zA-Z]+/d{4}", "<your replacement>"
    )


def create_line_scrubber(remove_lines_containing: str) -> Scrubber:
    return lambda t: scrub_lines_containing(remove_lines_containing, t)


def scrub_lines_containing(remove_lines_containing: str, text: str) -> str:
    lines = text.splitlines()
    for line in lines:
        if remove_lines_containing in line:
            lines.remove(line)
    return "\n".join(lines)
