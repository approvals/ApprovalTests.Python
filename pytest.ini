[pytest]
testpaths = tests
addopts = --showlocals

asyncio_default_fixture_loop_scope = function
