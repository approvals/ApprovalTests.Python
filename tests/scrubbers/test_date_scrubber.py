from pytest import CaptureFixture

from approval_utilities.utilities import markdown_table
from approvaltests import Options, verify, verify_exception
from approvaltests.scrubbers.date_scrubber import DateScrubber


def test_supported_formats() -> None:
    supported_formats = DateScrubber.get_supported_formats()
    for date_regex, examples in supported_formats:
        for example in examples:
            assert DateScrubber(date_regex).scrub(example) == "<date0>"


def test_supported_formats_arbitrary_string() -> None:
    assert (
        DateScrubber("[a-zA-Z]{3} [a-zA-Z]{3} \\d{2} \\d{2}:\\d{2}:\\d{2}").scrub(
            "arbitrary string"
        )
        == "arbitrary string"
    )


def test_supported_format_example() -> None:
    # begin-snippet: scrub-date-example
    verify(
        "created at 03:14:15",
        options=Options().with_scrubber(DateScrubber.get_scrubber_for("00:00:00")),
    )
    # end-snippet


def test_adds_valid_date_scrubber() -> None:
    """
    Hello on <date0>
    """
    text = "02025-007-020"
    try:
        DateScrubber.add_scrubber(text, r"\d{5}-\d{3}-\d{3}", display_message=False)
        verify(
            f"Hello on {text}",
            options=Options()
            .with_scrubber(DateScrubber.get_scrubber_for(text))
            .inline(),
        )
    finally:
        DateScrubber._clear_custom_scrubbers()  # Clean up after test, even if it fails


def test_raises_error_if_regex_does_not_match_example() -> None:
    """
    Exception: Regex '\d{2}/\d{2}/\d{4}' does not match example '2025-07-20'
    """

    def call() -> None:
        DateScrubber.add_scrubber("2025-07-20", r"\d{2}/\d{2}/\d{4}")

    verify_exception(call, options=Options().inline())


def test_handles_invalid_regex_patterns_gracefully() -> None:
    """
    Exception: Invalid regex pattern '[invalid-regex': unterminated character set at position 0
    """

    def call() -> None:
        DateScrubber.add_scrubber("2025-07-20", "[invalid-regex")

    verify_exception(call, options=Options().inline())


def test_unsupported_format() -> None:
    verify_exception(
        lambda: DateScrubber.get_scrubber_for("AN_UNSUPPORTED_DATE_FORMAT")
    )


def test_custom_date_format_example() -> None:
    # begin-snippet: custom_date_format_example
    from approvaltests.scrubbers.date_scrubber import DateScrubber

    # Add a custom date format
    DateScrubber.add_scrubber("2025-07-20", r"\d{4}-\d{2}-\d{2}", display_message=False)

    # Now you can use it in your tests
    verify(
        "Event scheduled for 2025-07-20",
        options=Options().with_scrubber(DateScrubber.get_scrubber_for("2025-07-20")),
    )
    # end-snippet
    # Clean up after test
    DateScrubber._clear_custom_scrubbers()


def test_supported_formats_as_table() -> None:
    table = markdown_table.MarkdownTable.with_headers("Example Date", "Regex Pattern")
    supported_formats = DateScrubber.get_supported_formats()
    for date_regex, examples in supported_formats:
        table.add_rows(examples[0], date_regex)
    verify(table)


def test_custom_scrubber_displays_message(capsys: CaptureFixture[str]) -> None:
    """
    You are using a custom date scrubber. If you think the format you want to scrub would be useful for others, please add it to https://github.com/approvals/ApprovalTests.Python/issues/124.

    To suppress this message, use
        DateScrubber.add_scrubber("2023-Dec-25", "\d{4}-[A-Za-z]{3}-\d{2}", display_message=False)

    """
    try:
        DateScrubber.add_scrubber("2023-Dec-25", r"\d{4}-[A-Za-z]{3}-\d{2}")
        captured = capsys.readouterr()
        verify(captured.out, options=Options().inline())
    finally:
        DateScrubber._clear_custom_scrubbers()


def test_custom_scrubber_message_can_be_suppressed(capsys: CaptureFixture[str]) -> None:
    """ """
    try:
        DateScrubber.add_scrubber(
            "2023-Dec-25", r"\d{4}-[A-Za-z]{3}-\d{2}", display_message=False
        )
        captured = capsys.readouterr()
        verify(captured.out, options=Options().inline())
    finally:
        DateScrubber._clear_custom_scrubbers()
