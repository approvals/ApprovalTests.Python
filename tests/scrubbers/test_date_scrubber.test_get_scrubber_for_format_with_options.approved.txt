Event logged at <date0> by system
